Add initial docs (#21)
dapr-sandbox/components-go-sdk: daprdocs/content/en/go-sdk-docs/_index.md

linkTitle: "Go"
weight: 1000
description: How to get up and running with the Dapr pluggable components Go SDK
no_list: true
is_preview: true
---

Dapr offers packages to help with the development of Go pluggable components.

## Prerequisites

- [Go 1.19](https://go.dev/dl/) or later
- [Dapr 1.9 CLI]({{< ref install-dapr-cli.md >}}) or later
- Initialized [Dapr environment]({{< ref install-dapr-selfhost.md >}})
- Linux, Mac, or Windows (with WSL)

{{% alert title="Note" color="primary" %}}
Development of Dapr pluggable components on Windows requires WSL as some development platforms do not fully support Unix Domain Sockets on "native" Windows.
{{% /alert %}}

## Application creation

Creating a pluggable component starts with an empty Go application.

```bash
mkdir example
cd component
go mod init example
```

## Import Dapr packages

Import the Dapr pluggable components SDK package.

```bash
go get github.com/dapr-sandbox/components-go-sdk@v0.1.0
```

## Create main package

In `main.go`, import the Dapr plugggable components package and run the application.

```go
package main

import (
	dapr "github.com/dapr-sandbox/components-go-sdk"
)

func main() {
	dapr.MustRun()
}
```

> This creates an application with no components; you will need to implement and register one or more components.

Author: Updated.

## Implement and register components

 - [Implementing an input/output binding component]({{< ref go-bindings >}})

>  - [Implementing a pub-sub component]({{< ref go-pub-sub >}})

Author: Updated.

 - [Implementing a state store component]({{< ref go-state-store >}})

{{% alert title="Note" color="primary" %}}
Only a single component of each type can be registered with an individual service. However, [multiple components of the same type can be spread across multiple services]({{< ref go-advanced >}}).
{{% /alert %}}

## Test components locally

Pluggable components can be tested by starting the application on the command line and configuring a Dapr sidecar to use it.

To start the component, in the application directory:

```bash
go run main.go
```

> To configure Dapr to use the (state store) component, in the resources path directory:

Author: Updated.

## Create container

Pluggable components are deployed as containers that run as sidecars to the application (like Dapr itself). A typical `Dockerfile` for creating a Docker image for a Go application might look like:

```dockerfile
FROM golang:1.20-alpine AS builder

WORKDIR /usr/src/app

# Download dependencies
COPY go.mod go.sum ./
RUN go mod download && go mod verify

# Build the application
COPY . .
RUN go build -v -o /usr/src/bin/app .

FROM alpine:latest

# Setup non-root user and permissions
RUN addgroup -S app && adduser -S app -G app
RUN mkdir /tmp/dapr-components-sockets && chown app /tmp/dapr-components-sockets

# Copy application to runtime image
COPY --from=builder --chown=app /usr/src/bin/app /app

USER app

CMD ["/app"]
```

Build the image:

```bash
docker build -f Dockerfile -t <image name>:<tag> .
```

{{% alert title="Note" color="primary" %}}
Paths for `COPY` operations in the `Dockerfile` are relative to the Docker context passed when building the image, while the Docker context itself will vary depending on the needs of the application being built. In the example above, the assumption is that the Docker context is the component application directory.

> {{% /alert %}}

Author: Yes.

---
type: docs
title: "Getting started with the Dapr pluggable components Go SDK"
linkTitle: "Go"
weight: 1000
description: How to get up and running with the Dapr pluggable components Go SDK
no_list: true
is_preview: true
---

Dapr offers packages to help with the development of Go pluggable components.

## Prerequisites

- [Go 1.19](https://go.dev/dl/) or later
- [Dapr 1.9 CLI]({{< ref install-dapr-cli.md >}}) or later
- Initialized [Dapr environment]({{< ref install-dapr-selfhost.md >}})
- Linux, Mac, or Windows (with WSL)

{{% alert title="Note" color="primary" %}}

> Development of Dapr pluggable components on Windows requires WSL as some development platforms do not fully support Unix Domain Sockets on "native" Windows.

Reviewer: Keeping for reference: for Java, UDS support is pending on https://github.com/netty/netty/issues/10991

> Development of Dapr pluggable components on Windows requires WSL as some development platforms do not fully support Unix Domain Sockets on "native" Windows.

Author: I changed to "Development of Dapr pluggable components on Windows requires WSL. Not all languages and SDKs expose Unix Domain Sockets on "native" Windows."

I feel saying "requires" is better than "recommend" as it just doesn't work today in "native" Windows.

> Development of Dapr pluggable components on Windows requires WSL as some development platforms do not fully support Unix Domain Sockets on "native" Windows.

Author: (I originally went with "some development platforms" rather than call out Go libraries as the limitation.)

{{% /alert %}}

## Application creation

Creating a pluggable component starts with an empty Go application.

```bash
mkdir example
cd component
go mod init example
```

## Import Dapr packages

Import the Dapr pluggable components SDK package.

```bash
go get github.com/dapr-sandbox/components-go-sdk@v0.1.0
```

## Create main package

In `main.go`, import the Dapr plugggable components package and run the application.

```go
package main

import (
	dapr "github.com/dapr-sandbox/components-go-sdk"
)

func main() {
	dapr.MustRun()
}
```

This creates an application with no components. You will need to implement and register one or more components.

## Implement and register components

 - [Implementing an input/output binding component]({{< ref go-bindings >}})
 - [Implementing a pub/sub component]({{< ref go-pub-sub >}})
 - [Implementing a state store component]({{< ref go-state-store >}})

{{% alert title="Note" color="primary" %}}
Only a single component of each type can be registered with an individual service. However, [multiple components of the same type can be spread across multiple services]({{< ref go-advanced >}}).
{{% /alert %}}

## Test components locally

> 
> Pluggable components can be tested by starting the application on the command line and configuring a Dapr sidecar to use it.
> 
> To start the component, in the application directory:

Reviewer: I think we need to mention the creation of the `/tmp/dapr-components-sockets` directory somewhere. This is required and without this, the steps described here will fail, no?

How about:
```suggestion

### Create a directory to hold your UDS

Dapr and your components will communicate using UDS files. By default, Dapr looks for those files in a directory named `/tmp/dapr-components-sockets`. If this directory doesn't exist, create it. Your component will also default to creating its Unix Domain Socket files in this diretory.

### Start your application

Pluggable components can be tested by starting the application on the command line and configuring a Dapr sidecar to use it.

To start the component, in the application directory:
```

```bash
go run main.go
```

> To configure Dapr to use the component, create a component YAML file in the resources directory. For example, for a state store component:

Reviewer: ```suggestion

### Configuring and running Dapr to use your pluggable component

To configure Dapr to use the component, create a component YAML file in the resources directory. For example, for a state store component:
```

Author: Split the "local testing" section into four subsections: creating temp directory, starting component, configuring Dapr, and starting Dapr (and application).

Creating a pluggable component starts with an empty Go application.

```bash
mkdir example
cd component
go mod init example
```

## Import Dapr packages

Import the Dapr pluggable components SDK package.

```bash
go get github.com/dapr-sandbox/components-go-sdk@v0.1.0
```

## Create main package

In `main.go`, import the Dapr plugggable components package and run the application.

```go
package main

import (
	dapr "github.com/dapr-sandbox/components-go-sdk"
)

func main() {
	dapr.MustRun()
}
```

This creates an application with no components. You will need to implement and register one or more components.

## Implement and register components

 - [Implementing an input/output binding component]({{< ref go-bindings >}})
 - [Implementing a pub/sub component]({{< ref go-pub-sub >}})
 - [Implementing a state store component]({{< ref go-state-store >}})

{{% alert title="Note" color="primary" %}}
Only a single component of each type can be registered with an individual service. However, [multiple components of the same type can be spread across multiple services]({{< ref go-advanced >}}).
{{% /alert %}}

## Test components locally

### Create the Dapr components socket directory

> Dapr communicates with pluggable components via Unix Domain Sockets files in a common directory. Both default to the `/tmp/dapr-components-sockets` directory. You should create this directory if it does not already exist.

Reviewer: "Both" sounded a big vague/ambiguous so I expended it a bit:

```suggestion
Dapr communicates with pluggable components via Unix Domain Sockets files in a common directory. By default, both Dapr and pluggable components will create and look for those UDS files in the `/tmp/dapr-components-sockets` directory. You should create this directory if it does not already exist.
```
